File: fastmq-common/src/main/java/com/song/fastmq/common/utils/Result.java
package com.song.fastmq.common.utils;

import java.util.concurrent.CountDownLatch;

/**
 * @author song
 */
public class Result<T> {

    private CountDownLatch latch = new CountDownLatch(1);

    private T data;

    private Throwable throwable;

    /**
     * Wait for async operation completes ,and then retrieves its result.
     *
     * @return the data
     * @throws Throwable
     */
    public T getData() throws Throwable {
        this.latch.await();
        return data;
    }

    /**
     * Only support setData once.
     *
     * @param data
     */
    public void setData(T data) {
        this.data = data;
        this.latch.countDown();
    }

    public Throwable getThrowable() {
        return throwable;
    }

    public void setThrowable(Throwable throwable) {
        this.latch.countDown();
        this.throwable = throwable;
    }
}

File: fastmq-common/src/main/java/com/song/fastmq/common/utils/Result.kt
package com.song.fastmq.common.utils

import java.util.concurrent.CountDownLatch

/**
 * @author song
 */
class Result<T> {

    private val latch = CountDownLatch(1)

    private var data: T? = null

    var throwable: Throwable? = null
        set(throwable) {
            this.latch.countDown()
            field = throwable
        }

    /**
     * Wait for async operation completes ,and then retrieves its result.
     *
     * @return the data
     * @throws Throwable
     */
    @Throws(Throwable::class)
    fun getData(): T? {
        this.latch.await()
        return data
    }

    /**
     * Only support setData once.
     *
     * @param data
     */
    fun setData(data: T) {
        this.data = data
        this.latch.countDown()
    }
}
